docs/CNDP_README.md

```diff
@@ -0,0 +1,164 @@
+
+Modify [pktgen_cndp.bess](conf/pktgen_cndp.bess) script as follows.
+
+1) Update source and destination interface mac addresses of the access and core interface  - smac_access, smac_core, dmac_access, dmac_core. Here smac_xxx corresponds to mac address of NIC in system 2 where we run pktgen and dst_xxx corresponds to mac address of NIC in system 1 which runs UPF pipeline.
+
+2) Update worker core ids to use core id in NUMA node where NIC is attached. For example if NIC is attached to NUMA node 1, use worker core ids in NUMA node 1. eg:  `workers=[22, 23, 24, 25]`
+
+3) Bind the NICs to DPDK and note the vfio device number in "/dev/vfio"
+
+From the top level directory call: (**Note**: Update below command to set `cpuset-cpus`range same as worker core ids in step 2 above and use the vfio device number from step 3)
+
+```
+docker run --name pktgen -td --restart unless-stopped \
+        --cpuset-cpus=22-25 --ulimit memlock=-1 --cap-add IPC_LOCK \
+        -v /dev/hugepages:/dev/hugepages -v "$PWD/conf":/opt/bess/bessctl/conf \
+        -v /lib/firmware/intel:/lib/firmware/intel \
+        --device=/dev/vfio/vfio --device=/dev/vfio/119 --device=/dev/vfio/120 \
+        upf-epc-bess:"$(<VERSION)" -grpc-url=0.0.0.0:10514
+
+docker exec -it pktgen ./bessctl run pktgen
+```
+
+We can monitor if pktgen is sending packets using the following command:
+
+```
+docker exec -it pktgen ./bessctl monitor tc
+```
+
+If we need to stop sending packets at some point use the following command:
+
+```
+docker exec -it pktgen ./bessctl daemon reset
+```
+
+### CNDP OMEC-UPF multiple worker threads setup
+
+Modify OMEC UPF and CNDP configuration files to support multiple worker threads. Each thread will run UPF pipeline in a different core.
+
+To test multiple worker thread, we need to use CNDP jsonc file with appropriate configuration. An example CNDP jsonc file for 4 worker threads is in [cndp_upf_4worker.jsonc](./conf/cndp_upf_4worker.jsonc). Follow below steps to configure OMEC-UPF pipeline using 4 worker threads which runs on 5 cores (4 BESS worker threads and 1 main thread).
+
+1) Update [upf.json](./conf/upf.json) to set number if worker threads as 4. `"workers": 4,`
+
+2) Update [upf.json](./conf/upf.json) to use appropriate CNDP jsonc file with required number of queues.
+
+`"cndp_jsonc_file": "/opt/bess/bessctl/conf/cndp_upf_4worker.jsonc"`
+
+3) Modify the script [docker_setup.sh](./scripts/docker_setup.sh) and update the function `move_ifaces()` in condition `if [ "$mode" == 'cndp' ]`.
+
+Update `num_q` value same as number of worker threads
+
+Update `start_q_idx` to choose the start queue index to receive n/w packets.
+
+4) Modify the script  [docker_setup.sh](./scripts/docker_setup.sh) to assign 5 cores (4 worker thread, 1 main thread) to run BESS UPF pipeline. `--cpuset-cpus=22-26`
+
+5) From the top level directory call:
+```
+$ ./scripts/cndp_reset_upf.sh
+$ ./scripts/docker_setup.sh
+```
```

Comment:
I prefer to keep CNDP related setup procedure in a separate file since there are lot of steps. I'll add `Configuration: CNDP` mode in [INSTALL.md](https://github.com/omec-project/upf/blob/master/docs/INSTALL.md) and then refer CNDP_README.md